File: tests/kotlin/02_log_redirect_deser.kt
// Kotlin Log Injection, Open Redirect, Deserialization, XPath, LDAP, Template Tests

fun main() {
    val userInput = System.getenv("INPUT") ?: ""
    val targetUrl = System.getenv("URL") ?: ""
    val jsonData = System.getenv("DATA") ?: ""
    val xpathExpr = System.getenv("XPATH") ?: ""
    val filter = System.getenv("FILTER") ?: ""
    val template = System.getenv("TEMPLATE") ?: ""

    // CVI_9307 - 日志注入
    logger.info("User request: $userInput")
    logger.warn("Warning: $userInput")
    logger.error("Error: $userInput")

    // CVI_9308 - 开放重定向
    sendRedirect(targetUrl)
    response.sendRedirect(targetUrl)

    // CVI_9309 - 反序列化
    val result = Gson().fromJson(jsonData, User::class.java)
    val obj = ObjectMapper().readValue(jsonData, Object::class.java)

    // CVI_9310 - XPath注入
    val xpath = XPathFactory.newInstance().newXPath()
    xpath.evaluate(xpathExpr, document)

    // CVI_9311 - LDAP注入
    val ctx = InitialLdapContext()
    val searchControls = SearchControls()

    // CVI_9312 - 模板注入
    val engine = TemplateEngine()
    engine.process(template, context)

    // False positives
    logger.info("Server started")
    sendRedirect("/dashboard")
    Gson().fromJson("{\"name\":\"test\"}", User::class.java)
    xpath.evaluate("//book/title", document)
    engine.process("hello", context)
}

File: tests/lua/02_log_redirect_deser.lua
-- Lua Open Redirect, Template Injection, Log Injection, XPath, Deserialization Tests

local userInput = os.getenv("INPUT") or ""
local targetUrl = os.getenv("URL") or ""
local jsonData = os.getenv("DATA") or ""

-- CVI_9607 - 开放重定向
ngx.redirect(targetUrl)
redirect(targetUrl)

-- CVI_9608 - 模板注入
local tmpl = template.compile(userInput)
template.render(userInput)
template.process(userInput)

-- CVI_9609 - 日志注入
log.info("User input: " .. userInput)
log.warn("Warning: " .. userInput)
log.error("Error: " .. userInput)

-- CVI_9610 - XPath注入
local result = xpath.parse(userInput)
xpath.selectNodes(userInput)

-- CVI_9611 - 不安全反序列化
local data = json.decode(jsonData)
local obj = cjson.decode(jsonData)
local msg = msgpack.unpack(jsonData)

-- False positives
ngx.redirect("/dashboard")
log.info("Server started")
json.decode("{}")
template.render("hello")
